src/collapse/xctrace.rs

```diff
@@ -0,0 +1,492 @@
+                        (Some(CurrentTag::Backtrace { .. }), FRAME) => {
+                            let (id, name) = attributes_to_frame(&attributes)?;
+                            Some(CurrentTag::Frame { id, name })
+                        }
+                        (Some(_), _) => {
+                            // Skip tag combination we are not interested in.
+                            // Tags' matchedness&validity have alreay been checked by quick_xml:
+                            // - https://docs.rs/quick-xml/latest/quick_xml/reader/struct.Config.html#structfield.allow_unmatched_ends
+                            // - https://docs.rs/quick-xml/latest/quick_xml/reader/struct.Config.html#structfield.check_end_names
+                            None
+                        }
+                    };
+                    if let Some(new_state) = new_state {
+                        context.state_backtrace.push_back(new_state);
+                    }
+                }
+                Event::End(end) => {
+                    let name = end.name().into_inner();
+                    // Skip unknown tags
+                    if !is_interested_tag(name) {
+                        continue;
+                    }
+                    let Some(state) = context.state_backtrace.pop_with_name(name) else {
+                        return invalid_data_error!(
+                            "Unpaired tag: {}",
+                            String::from_utf8_lossy(name)
+                        );
+                    };
+                    // Retrieve information when a tag span ends.
+                    match (context.state_backtrace.top_mut(), state) {
+                        (None, CurrentTag::TraceQueryResult { nodes }) => {
+                            break nodes;
+                        }
+                        (
+                            Some(CurrentTag::TraceQueryResult { nodes }),
+                            CurrentTag::Node { rows },
+                        ) => {
+                            nodes.push(Node { rows });
+                        }
+                        (Some(CurrentTag::Node { rows }), CurrentTag::Row { backtrace }) => {
+                            // <backtrace/> in some row is replaced with <sentinel/>, hence we ignore thess rows.
+                            if let Some(backtrace) = backtrace {
+                                rows.push(Row { backtrace });
+                            }
+                        }
+                        (
+                            Some(CurrentTag::Row { backtrace }),
+                            CurrentTag::Backtrace { id, frames },
+                        ) => {
+                            let new_backtrace = Arc::new(Backtrace { id, frames });
+                            let ret = context
+                                .backtraces
+                                .insert(new_backtrace.id, new_backtrace.clone());
+                            assert!(ret.is_none());
```

Comment:
```suggestion
                            assert_eq!(ret, None);
```
Gives better errors if the assetion fails.
Although I wonder if this should instead return an `Err` saying "two backtraces with the same ID encountered"?

Reply:
Fixed